src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -24,15 +35,60 @@ public synchronized Product getProduct(@Nullable Long id) throws ProductNotFound
         return products.stream()
                 .filter(product -> id.equals(product.getId()))
                 .findAny()
-                .orElseThrow(ProductNotFoundException::new);
+                .orElseThrow(() -> new ProductNotFoundException(id));
     }
 
     @Override
-    public synchronized List<Product> findProducts() {
-        return products.stream()
+    public synchronized List<Product> findProducts(String query, SortField sortField, SortOrder sortOrder) {
+        List<Product> processingProducts = products.stream()
                 .filter(product -> product.getPrice() != null)
                 .filter(product -> product.getStock() > 0)
                 .collect(Collectors.toList());
+
+        if (query != null && !query.isEmpty()) {
+            processingProducts = getQueriedProducts(processingProducts, query);
+        }
+
+        if (sortField != null && sortOrder != null) {
+            processingProducts = getSortedProducts(processingProducts, sortField, sortOrder);
+        }
```

Comment:
Please refactor to avoid twice sorting in the case when you have both: query and some specific sort